### ISSUE TITLE
Normalizing error checking

### ISSUE THREAD

**Issue by danpovey** [k2-fsa/k2#111]
Guys, the error-checking stuff is still quite confused.
There is a file error.h that probably shouldn't be there, and glog is included by both error.h and context.h.

My proposal is to limit ourselves to the macros in debug.h and not use glog, because it's too complex and
IMO not really necessary.  Any feedback on this?

**Comment by qindazhu** [comment]
`error.h` contains only one function `CheckCudaError`, I created it when @songmeixu did't committ PR about `debug.h`, now the whole file can be deleted (I'll do this now).

RE glog, I'm OK with removing glog, but for now, we have no `CHECK` marocs with printing out messages, e.g. `k2_CHECK_EQ() << "message"`, and Log level mechanism, such as `LOG(FATAL)<<`, maybe @songmeixu would like to add those code?

**Comment by songmeixu** [comment]
IE, the debug.h is expected to be used for cuda codes. Where the print msg would be silenced finally for release. Thus it's not for logging, but for debugging (in both DEBUG/NDEBUG mode, as `K2_PARANOID_ASSERT` takes effect in both).

----

As to non-cuda codes, I think msg is needed for logging and debugging/error-handling, thus glog/loguru is sth that valued. Consider the pytorch/tf is the first target for K2 to work with. They both utilized glog, I would vote glog +1. The problem to dev with glog in a library (K2 would be at 1st stage?) is, without a config which must be the final `main()` job, there should be a local one with the same macro-name. This is a convention to use with glog at least for a library. During called through `main()`, the preset define-check macro(`#ifndef glog..`) is used to choose from the real glog or the mimic one. In this way, the glog can be able to different mother-ships.

**Comment by songmeixu** [comment]
A library should not expect a macro "the-predefined-glog-indicator-in-glog-header". If glog is not presented, the loguru should be ready, that means be inited in a default way in its part of the glog-or-loguru header. I made it init as so. Would check it to make sure it works (write tests).

**Comment by songmeixu** [comment]
Another way is forget glog, define our one. But the default msg distributions (stdout/stderr) may make the upper-framework unhappy. That's a reason why glog is used widely.

**Comment by danpovey** [comment]
Meixu, I found your comment ". The problem to dev with glog in a library (K2 would be at 1st stage?) is, without a config which must be the final main() job, there should be a local one with the same macro-name. This is a convention to use with glog at least for a library." confusing.  Can you please clarify with example code?

**Comment by songmeixu** [comment]
Sorry about it and sure. I think the right way is to introduce loguru. And would write tests to show.

**Comment by qindazhu** [comment]
If we have to use third-party log libraries, I would like to use glog instead of loguru. As you said before, a lot of projects  use glog (e.g. PyTorch, TensorFlow). Or we write our own log system as we only need a few features (e.g. `CHECK` with printing error message, `LOG(FATAL`))


> Sorry about it and sure. I think the right way is to introduce loguru. And would write tests to show.

**Comment by songmeixu** [comment]
How do I feel like no one knows what I'm saying 🐶🐶🐶 I can't express myself well in Eng. If there is glog, there must be another glog that mimics it.

**Comment by qindazhu** [comment]
OK, sorry, you mean there must be some init function to call InitGoogleGlog? or can you explain the requirements/limitation you declared with some sample code?

**Comment by songmeixu** [comment]
sure, as dan has asked about.
